src/util/domains.py

```diff
@@ -45,6 +43,95 @@
 except NameError:
     basestring = unicode = str
 
+# import from domains/expression
+try:
+    import odoo.domains as _dom
+
+    FALSE_LEAF = _dom._FALSE_LEAF
+    TRUE_LEAF = _dom._TRUE_LEAF
+    NOT_OPERATOR = _dom.DomainNot.OPERATOR
+    AND_OPERATOR = _dom.DomainAnd.OPERATOR
+    OR_OPERATOR = _dom.DomainOr.OPERATOR
+    DOMAIN_OPERATORS = {NOT_OPERATOR, AND_OPERATOR, OR_OPERATOR}
+
+    # normalization functions are redefined here because they will be deprecated
+    # the Domain factory normalized but also distributes operators during creation of domains
+
+    def normalize_domain(domain):
+        """Return a normalized version of the domain.
+
+        The normalized version have all implicit '&' operators explicit.
+        One property of normalized domain expressions is that they
+        can be easily combined together as if they were single domain components.
+        """
+        assert isinstance(
+            domain, (list, tuple)
+        ), "Domains to normalize must have a 'domain' form: a list or tuple of domain components"
+        if not domain:
+            return [TRUE_LEAF]
+        result = []
+        expected = 1
+        op_arity = {NOT_OPERATOR: 1, AND_OPERATOR: 2, OR_OPERATOR: 2}
+        for token in domain:
+            if expected == 0:  # more than expected, like in [A, B]
+                result[0:0] = [AND_OPERATOR]  # put an extra '&' in front
+                expected = 1
+            if isinstance(token, (list, tuple)):  # domain term
+                expected -= 1
+                if len(token) == 3 and token[1] in ("any", "not any"):
+                    new_token = (token[0], token[1], normalize_domain(token[2]))
+                    result.append(new_token)
+                else:
+                    result.append(normalize_leaf(token))
+            else:
+                expected += op_arity.get(token, 0) - 1
+                result.append(token)
+        if expected:
+            raise ValueError("Domain {!r} is syntactically not correct.".format(domain))
+        return result
+
+    def normalize_leaf(leaf):
+        if not is_leaf(leaf):
+            raise TypeError("Leaf must be a tuple or list of 3 values")
+        left, operator, right = leaf
+        original = operator
+        operator = operator.lower()
+        if operator == "<>":
+            operator = "!="
+        if isinstance(right, bool) and operator in ("in", "not in"):
+            _logger.warning("The domain term '%s' should use the '=' or '!=' operator.", ((left, original, right),))
```

Comment:
Is this still true?
```py
>>> list(domains.Domain("id","in",True)._optimize(self.env['base']))
[('id', 'in', [True])]
>>> list(domains.Domain("active","in",True)._optimize(self.env['res.users']))
[('active', 'in', [True])]
>>> list(domains.Domain("active","in",False)._optimize(self.env['res.users']))
[(0, '=', 1)]
```
The last output looks odd... I'd expect an error or `[('active','in',[False])]`/`[('active','=',False)]` as output.

If we no longer show any warning then why would we show one during the upgrade. If we now silently replace the leaf by something else, the warnings should be now more like `This is translated to ... since Odoo 19`.

I still think it's better to implement `normalize_leaf` using the new machinery at least for the cases where it doesn't do anything too involved.

Reply:
> The implementation is a copy of what is done today.

That's what I understood. My point: we cannot keep for the future what's done today. It should be aligned. For merging the task this may be OK, but for the long term whatever we do here should be aligned with what's expected from the implementation in community.

Reply:
Well spotted.